Reversed Range and Range.__getitem__

smichr opened the pull request:
closes #10660 
fixes #10606
- [x] hand check tests

asmeurer committed:
Disallow step size 0 in Range

Previously it was converted to 1 due to logic that was guarding against None
from the slice.

asmeurer committed:
Make reversed Ranges keep their order

Fixes #10606.

The logic in Range._intersect for intersecting with intervals is particularly
nasty, because of the corner cases with infinite ranges (ranges with infinite
starts are handled differently). I believe the test cases test all the proper
combinations, but the logic should still probably be checked, and it likely
can be simplified.

This changes the behavior of Range.__iter__ to raise an exception for ranges
with infinite starts, since ranges always iterate in order. See the discussion
on the issue for more information about this. I felt that it was better to
always iterate in order, and catch possible bugs if the range is reversed,
than to iterate backwards in this case. It's worth noting, however, that this
makes it more annoying to determine the values in an infinite range beyond the
two that are printed by the pretty printer. But to fix this, I think we should
add better support for slicing ranges.

Also added a better docstring to Range.

asmeurer committed:
Fix pretty and latex printing of Range

It was doing the wrong thing for infinite and reversed ranges.

asmeurer committed:
Small fixes to Range docstring

asmeurer committed:
Delete unused lines

asmeurer committed:
Add some additional test cases for Range.intersect(Interval)

smichr commented:
ping @asmeurer

smichr commented:
Next time I won't write the `__getitem__`, either ;-)

AnishShah commented on sympy/sets/fancysets.py:
```diff
@@ -334,66 +385,83 @@ def __new__(cls, *args):
-            return S.EmptySet
+        if all(i.is_infinite for i in  (start, stop)):
+            raise ValueError("The start and end value of "
+                             "Range cannot both be unbounded")
```
Since we have deprecated `bounded/unbounded` in favour of `finite/infinite` in assumptions, should we change it to `infinite` here? So that we can maintain same things in all places? 
Just a suggestion. Let me know what you think.

smichr commented:
I'll commit this this evening if there are no more comments. Do you agree that re-striding from the infinite end is a bad idea, @asmeurer ? We grant that `Range(oo, end, 1)` will always end at `end - 1` but if you then slice with `[::2]` it is ambiguous which track you will be on, odds or evens.

asmeurer commented:
Wow, there's no way I can review that `__getitem__`. I'll take your word that you tested all the cases. +1 to merge this if you feel the same way.

Ranges with infinite start seem like a bad idea in the sense that they make this whole thing more complicated. I'm not quite sure if `Range(-oo, 0, 1)[::2]` is ambiguous. I'll have to think about it. 

And I don't know about you, but this whole exercise has convinced me that 1-based indexing is superior to 0-based indexing.

asmeurer commented on sympy/sets/fancysets.py:
```diff
@@ -404,52 +471,178 @@ def _intersect(self, other):
+    def __getitem__(self, i):
+        from sympy.functions.elementary.complexes import sign
+        from sympy.functions.elementary.integers import ceiling, floor
+        ooslice = "cannot slice from infinite end"
```
infinite start?

smichr replied:
The infinity can be at either end (Range(-oo, 0) or Range(0, oo)) so I left the message generic.

asmeurer replied:
Oh I see, the issue is slicing from the endpoint that has the infinity. There's probably a better term here that doesn't imply only the second argument of the range.

asmeurer commented on sympy/sets/fancysets.py:
```diff
@@ -404,52 +471,178 @@ def _intersect(self, other):
+        if isinstance(i, slice):
+            if self.size.is_finite:
+                start, stop, step = i.indices(self.size)
+                n = ceiling((stop - start)/step)
```
`//`?

smichr replied:
That ambiguity is the odd/even case that I mentioned above. We already agree that an infinite start has a blessed ending point but if you step through this from the infinite end, there is ambiguity as to which sequence (mod step) you will be on.

smichr replied:
Oh, I guess you mean by `//` that perhaps we should be using integer division. But then ceiling would not be necessary. In this file, `division` has been imported so we are truly moving `n` up. It could be done using divmod, too. What is here works, however, and is transparent in terms of what is being done.

asmeurer replied:
I wasn't sure what this represented, just checking that this wasn't a typo, since this looks very similar to `__len__`, which uses `//`.

asmeurer commented on sympy/sets/fancysets.py:
```diff
@@ -404,52 +471,178 @@ def _intersect(self, other):
+                    raise ValueError(zerostep)
+                step = i.step or 1
+                ss = step*self.step
+                model = list(range(*i.indices(1 +
```
I would have hoped we could do this without creating explicit lists. What if the resulting range is large?

smichr replied:
Ah yes. You are right. I've already enumerated the possibilities so will just put that there and remove this.

asmeurer commented:
That's about all the reviewing I can muster for `__getitem__`.

smichr commented:
> make this whole thing more complicated

Just having an infinite range is what complicates this for anything but a slice with no None in it. As you can see from the code, the infinity at one end can be transformed easily to the other but there are still *a lot of cases*: start, stop and end in {neg, None, 0, pos} = 64 possibilities.

smichr commented on sympy/sets/tests/test_fancysets.py:
```diff
@@ -140,20 +151,74 @@ def test_Range():
+    # test slicing
+    raises(ValueError, lambda: Range(2, -oo, -2)[2:2:0])
+    assert Range(2, -oo, -2)[2:2:2] is S.EmptySet
+    assert Range(2, -oo, -2)[:2:2] == Range(0, 4, 4)
```
- [ ] check that `Range(2,-oo,-2)[:4:2] == Range(2, -2, -2)`; the way this test is written makes it look like th estep will be positive

smichr commented:
@asmeurer, you mentioned that you didn't want to return `FiniteSet(single_element)` for a Range that would only return a single element. But if that single element is oo we must either a) raise an error, b) return {oo} or c) special case Range(oo, oo). If we choose (c) then we might as well special case Range(0, 0) to return that instead of EmptySet. I did (b) in the 2nd to last commit and (c) in the last commit.

Let me know which you (or anyone else reading this) prefers.

smichr commented:
@asmeurer , you mentioned that you didn't want to see a singleton range like `Range(0, 1)` become `FiniteSet(1)`. So I've allowed created Range(oo, oo, 1) -> to represent a singleton range containing an infinite value as would be returned from something like `Range(oo)[-1:]`. I've also represented the null range as `Range(0, 0, 0)` rather than returning an EmptySet; the implications of that are that EmptySet will never be returned from an intersection with another set `Range(foo).intersect(bar)` -> `Range(...)`. This could create problems when someone does a set intersection and checks `if intersection is S.EmptySet` rather than `if intersection`. But is it really a problem if in 2 cases a FiniteSet is returned -- {oo} and {-oo} and EmptySet is returned instead of Range(0,0,0)? The FiniteSet and EmptySet are both iterable; the FiniteSet has inf and sup values (though the EmptySet does not). I am not sure of how these will matter in code generation -- your use-case rationale for not wanting non-Range values. Could you advise?

asmeurer commented:
Wouldn't Range(oo, oo) be empty, because of the way ranges don't include their upper bound (oo < oo is False)?

asmeurer commented:
I guess it's 6 of one, half dozen of the other for Range(0, 0, 1) vs. EmptySet. If we return the former, you have to be careful in checks against EmptySets (maybe Set ought to have an `is_empty` property). If we return EmptySet, we have to be careful when we create a Range object (say, in code generation) and not assume that it will remain a range (like `r = Range(...); if not r == EmptySet: ...`).

asmeurer commented:
Also I thought infinity in both the start and end bounds of Range was disallowed in the constructor anyway.

smichr commented:
I'll go with 2 special cases ({oo}/{-oo} and {}) rather than create non-traditional Ranges.

asmeurer commented:
FWIW, range(0) is not really a non-traditional range (Python's `range` allows it just fine).

smichr committed:
implement slicing of Range

smichr commented:
> FWIW, range(0) is not really a non-traditional range

Just saw that this morning and came to the issue to report the same:

```
>>> range(1,0,10)
xrange(1, 1, 10)
```

So a null range is always `xrange(1, 1, step)`. As this is allowed and simply represents `[]` I think it makes sense to just change `Range(0, 0, 0)` to `Range(1, 1, step)`. For the infinite singleton...hmm. Perhaps `Range(0, oo, oo)` or `Range(0, -oo, -oo)` so it doesn't look like a null range?

smichr commented:
I'm not sure why the step is retained. Perhaps we should just always make the null Range be `Range(1, 1, 1)`.

asmeurer commented:
In Python 3 `range` just keeps whatever args you pass to it. I think the most reasonable default empty range should be `Range(0)` (i.e., `Range(0, 0, 1)`, the default start and step if you don't provide them are 0 and 1).

I'm still not really sold on any range giving an infinite singleton. I think we should disallow infinite step, and both start and stop being infinite. `Range(0, oo, oo)` just looks ambiguous as heck.

smichr commented:
> most reasonable default empty range should be Range(0)

+1

Regarding the infinite range -- I think it makes most sense to treat this like a subset of Integers and simply not include `oo`.

smichr committed:
Range(0) -> empty set; Range.intersect(Range)

Intersection.reduce short circuits on EmptySet() so
Range(oo).intersect(S.EmptySet) -> S.EmptySet and happens
outside of the Range methods. So rather than sometimes returning
Range(0) and sometimes EmptySet, all Range intersections that
are empty are returned as EmptySet. Slicing of the Range, however,
can return Range(0) (which, unlike xrange/range, does not retain
any information about how the empty Range was obtained).

Range/Range intersections are now defined.

Note about infinite values.
---------------------------

Although selection of an infinite value
in a Range is allowed, e.g. Range(oo)[-1] -> oo, slicing done from
the end containing the infinite value is disallowed since

1) there is no way to count the nth value from oo, e.g. Range(oo)[-2:]
2) there is no way to represent a single infinite value while
retaining normal Range semantics, Range(oo, oo + 1, 1) -> Range(oo, oo, 1)
which would be the empty set. So although Range(oo)[-1:] could be
represented as FiniteSet(oo) or an infinite step could be used and
defined to give only the first value in a Range, e.g. Range(3,oo,oo) ->
Range(3, 4, 1) hence list(Range(oo, 0, -oo)) -> [oo], neither of these
seems like a good idea.

As oo in S.Integers is False, oo in Range(oo) is also False; there
is a slight iconsistency in letting Range(oo)[-1] give oo even though
oo is not in the Range, but I don't see a reason to complicate the
handling of inspection of the Range start/stop values.

smichr commented:
@asmeurer , I am done with this. It is 100% tested. If you could read the last commit message and see if that makes sense and you are in agreement, then this is ready to go.

asmeurer commented:
+1 to disallowing slicing at an infinite end.

asmeurer commented:
So I does this superceed https://github.com/sympy/sympy/pull/10660? I had thought that it was built on top of it but I just see two commits now. 

At any rate, feel free to merge if you are done.

asmeurer commented:
And thanks for your work on this. If you're looking for something to do after this I could still use a review on https://github.com/sympy/sympy/pull/10271 (and of course if you want me to review any of your other PRs let me know).

smichr commented:
> I had thought that it was built on top

Sorry, I can try to fix that. After the vers 1.0 I used the diff to recover a clean history.

smichr committed:
Merge branch 'master' into wip

Conflicts:
	sympy/printing/pretty/tests/test_pretty.py

smichr committed:
Merge branch 'master' into reversed-range

smichr committed:
Merge branch 'wip' into reversed-range

Conflicts:
	sympy/printing/latex.py
	sympy/printing/pretty/pretty.py
	sympy/sets/fancysets.py
	sympy/sets/tests/test_fancysets.py

smichr commented:
Glad this one is done. I'll take a look at the other PR of yours in the next few days.

This is ready to merge if tests pass.

smichr committed:
edit Range docstring
